File: api/routes/transactions.py
from fastapi import APIRouter, Depends, HTTPException

from api.auth import get_current_user, get_current_user_optional
from api.db import get_transactions, get_summary, reverse_transaction

router = APIRouter(tags=["transactions"])


@router.get("/transactions")
def list_transactions(
    benefit_id: str | None = None,
    tax_category: str | None = None,
    status: str | None = None,
    current_user: dict | None = Depends(get_current_user_optional),
):
    if not current_user:
        return {"transactions": []}
    return {"transactions": get_transactions(
        current_user["id"],
        benefit_id=benefit_id,
        tax_category=tax_category,
        status=status,
    )}


@router.get("/transactions/summary")
def transaction_summary(current_user: dict | None = Depends(get_current_user_optional)):
    if not current_user:
        return {"by_category": [], "total_applied": {"count": 0, "total_deductible": 0}}
    return get_summary(current_user["id"])


@router.delete("/transactions/{txn_id}")
def reverse_txn(txn_id: str, current_user: dict = Depends(get_current_user)):
    if not reverse_transaction(txn_id, current_user["id"]):
        raise HTTPException(404, f"Transaction '{txn_id}' not found")
    return {"reversed": True, "id": txn_id}

File: backend-ts/src/routes/transactions.ts
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { AppError } from "../lib/errors";
import { getSummary, getTransactions, reverseTransaction } from "../db/transactionsRepo";

const TransactionsQuerySchema = z.object({
  benefit_id: z.string().optional(),
  tax_category: z.string().optional(),
  status: z.string().optional()
});

const TransactionParamsSchema = z.object({
  txnId: z.string().min(1)
});

export async function registerTransactionsRoutes(app: FastifyInstance): Promise<void> {
  app.get("/transactions", { preHandler: app.authenticateOptional }, async (request) => {
    const user = request.currentUser;
    if (!user) {
      return { transactions: [] };
    }

    const query = TransactionsQuerySchema.parse(request.query ?? {});

    return {
      transactions: getTransactions(user.id, {
        benefitId: query.benefit_id,
        taxCategory: query.tax_category,
        status: query.status
      })
    };
  });

  app.get("/transactions/summary", { preHandler: app.authenticateOptional }, async (request) => {
    const user = request.currentUser;
    if (!user) {
      return {
        by_category: [],
        total_applied: {
          count: 0,
          total_deductible: 0
        }
      };
    }

    return getSummary(user.id);
  });

  app.delete("/transactions/:txnId", { preHandler: app.authenticate }, async (request) => {
    const user = request.currentUser;
    if (!user) {
      throw new AppError(401, "User not found");
    }

    const { txnId } = TransactionParamsSchema.parse(request.params ?? {});
    const reversed = reverseTransaction(txnId, user.id);
    if (!reversed) {
      throw new AppError(404, `Transaction '${txnId}' not found`);
    }

    return {
      reversed: true,
      id: txnId
    };
  });
}
